File: eco-core/core-plugin/src/main/java/com/willfp/ecoarmor/EcoArmorPlugin.java
package com.willfp.ecoarmor;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.eco.core.integrations.IntegrationLoader;
import com.willfp.ecoarmor.commands.CommandEcoarmor;
import com.willfp.ecoarmor.config.EcoArmorYml;
import com.willfp.ecoarmor.display.ArmorDisplay;
import com.willfp.ecoarmor.sets.ArmorSets;
import com.willfp.ecoarmor.sets.util.ArmorUtils;
import com.willfp.ecoarmor.sets.EffectiveDurabilityListener;
import com.willfp.ecoarmor.sets.PreventSkullPlaceListener;
import com.willfp.ecoarmor.upgrades.Tiers;
import com.willfp.ecoarmor.upgrades.AdvancementShardListener;
import com.willfp.ecoarmor.upgrades.CrystalListener;
import com.willfp.ecoarmor.util.DiscoverRecipeListener;
import com.willfp.ecoarmor.util.EffectListener;
import com.willfp.libreforge.Holder;
import com.willfp.libreforge.LibReforge;
import lombok.Getter;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@SuppressWarnings("unused")
public class EcoArmorPlugin extends EcoPlugin {
    /**
     * Instance of EcoArmor.
     */
    @Getter
    private static EcoArmorPlugin instance;

    /**
     * ecoarmor.yml.
     */
    @Getter
    private final EcoArmorYml ecoArmorYml;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public EcoArmorPlugin() {
        super(687, 10002, "&c");
        instance = this;

        this.ecoArmorYml = new EcoArmorYml(this);

        LibReforge.init(this);
        LibReforge.registerHolderProvider(player -> {
            Holder active = ArmorUtils.getActiveSet(player);
            if (active == null) {
                return Collections.emptyList();
            } else {
                return Collections.singletonList(active);
            }
        });
    }

    @Override
    protected void handleEnable() {
        LibReforge.enable(this);
    }

    @Override
    protected void handleDisable() {
        LibReforge.disable(this);
    }

    @Override
    protected void handleReload() {
        this.getLogger().info(Tiers.values().size() + " Tiers Loaded");
        this.getLogger().info(ArmorSets.values().size() + " Sets Loaded");
        LibReforge.reload(this);
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandEcoarmor(this)
        );
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new CrystalListener(this),
                new AdvancementShardListener(this),
                new EffectiveDurabilityListener(this),
                new DiscoverRecipeListener(this),
                new PreventSkullPlaceListener(),
                new EffectListener()
        );
    }

    @Override
    protected List<IntegrationLoader> loadIntegrationLoaders() {
        return LibReforge.getIntegrationLoaders();
    }

    @Override
    protected @Nullable DisplayModule createDisplayModule() {
        return new ArmorDisplay(this);
    }

    @Override
    public String getMinimumEcoVersion() {
        return "6.15.0";
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoarmor/EcoArmorPlugin.kt
package com.willfp.ecoarmor

import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.items.Items
import com.willfp.ecoarmor.commands.CommandEcoArmor
import com.willfp.ecoarmor.display.ArmorDisplay
import com.willfp.ecoarmor.libreforge.ConditionIsWearingSet
import com.willfp.ecoarmor.sets.ArmorSetEquipSoundListeners
import com.willfp.ecoarmor.sets.ArmorSets
import com.willfp.ecoarmor.sets.ArmorUtils
import com.willfp.ecoarmor.sets.EffectiveDurabilityListener
import com.willfp.ecoarmor.sets.PlayerArmorSetEventListeners
import com.willfp.ecoarmor.sets.PreventSkullPlaceListener
import com.willfp.ecoarmor.upgrades.AdvancementShardListener
import com.willfp.ecoarmor.upgrades.CrystalListener
import com.willfp.ecoarmor.upgrades.TierArgParser
import com.willfp.ecoarmor.upgrades.Tiers
import com.willfp.ecoarmor.util.DiscoverRecipeListener
import com.willfp.libreforge.conditions.Conditions
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.registerSpecificHolderProvider
import org.bukkit.entity.LivingEntity
import org.bukkit.event.Listener

internal lateinit var plugin: EcoArmorPlugin
    private set

class EcoArmorPlugin : LibreforgePlugin() {
    init {
        plugin = this
        Items.registerArgParser(TierArgParser())
    }

    override fun handleLoad() {
        Conditions.register(ConditionIsWearingSet)
    }

    override fun handleEnable() {
        registerSpecificHolderProvider<LivingEntity> {
            ArmorUtils.getActiveHolders(it)
        }
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            Tiers,
            ArmorSets
        )
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandEcoArmor
        )
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            CrystalListener,
            AdvancementShardListener,
            EffectiveDurabilityListener,
            DiscoverRecipeListener,
            PreventSkullPlaceListener,
            ArmorSetEquipSoundListeners
        )
    }

    @Suppress("OVERRIDE_DEPRECATION")
    override fun createDisplayModule(): DisplayModule {
        return ArmorDisplay
    }
}
